base/abstractarray.jl

```diff
@@ -687,6 +688,10 @@ julia> empty([1.0, 2.0, 3.0], String)
 ```
 """
 empty(a::AbstractVector{T}, ::Type{U}=T) where {T,U} = Vector{U}()
+function empty(a::Type{<:AbstractArray{T,N}}, ::Type{U}=T) where {T,N,U}
+    convert(AbstractArray{U}, a(undef, Tuple(Iterators.repeated(0,N))))
```

Comment:
Unfortunately I'm not sure `T(undef, dims)` is a required part of the `AbstractArray` API. The operation of creating a new array from a type but with a different element type could potentially be handled by `similar` (see https://github.com/JuliaLang/julia/pull/20815).

Also this should be faster:
```suggestion
    convert(AbstractArray{U}, a(undef, ntuple(_ -> 0, N)))
```
BTW, better use the short function syntax as for other definitions.

Reply:
Thanks for the `ntuple` correction! I actually found that function some days ago and was gonna modify the PR accordingly but you were faster. :)

About the `T(undef, dims)`, yeah, sorry; that was stupid since `similar` has the `dims` argument. I completely forgot about it.
I also love your referenced PR; just today I ran into a case where it would have been useful.

> BTW, better use the short function syntax as for other definitions.

Do you mean I should always use `empty(x) = ...`? I didn't use it as the functions were getting too wide (over 92 columns) and I thought the short syntax should be avoided when the function spans multiple lines.

Reply:
> About the `T(undef, dims)`, yeah, sorry; that was stupid since `similar` has the `dims` argument. I completely forgot about it.
> I also love your referenced PR; just today I ran into a case where it would have been useful.

As I noted, the problem with `similar` is that it requires an instance, not a type.

> Do you mean I should always use `empty(x) = ...`? I didn't use it as the functions were getting too wide (over 92 columns) and I thought the short syntax should be avoided when the function spans multiple lines.

Yes it's fine to break after `=` (see examples in this file).

Reply:
> As I noted, the problem with similar is that it requires an instance, not a type.

_That_ was the problem with it, I remember... So first we will have to keep the constructor by type, if I'm seeing this correctly? Until #20815 is merged.

> Yes it's fine to break after = (see examples in this file).

I was thinking more in an aesthetic/readability sense. Only because it has been done in the past does not imply it has to be done again. But sure, I will change it!

Reply:
Actually I'm not even sure calling the constructor like that is guaranteed to work. Somebody else will have to confirm.